### PR TITLE
feature(shields): Add nice!view

### PR DESCRIPTION
**Pull request by Nicell**
## Board/Shield Check-list
 - [x] This board/shield is tested working on real hardware
 - [x] Definitions follow the general style of other shields/boards upstream ([Reference](https://zmk.dev/docs/development/new-shield))
 - [x] `.zmk.yml` metadata file added
 - [x] Proper Copyright + License headers added to applicable files (Generally, we stick to "The ZMK Contributors" for copyrights to help avoid churn when files get edited)
 - [x] General consistent formatting of DeviceTree files
 - [ ] Keymaps do not use deprecated key defines (Check using the [upgrader tool](https://zmk.dev/docs/codes/keymap-upgrader))
 - [x] `&pro_micro` used in favor of `&pro_micro_d/a` if applicable
 - [ ] If split, no name added for the right/peripheral half
 - [x] Kconfig.defconfig file correctly wraps *all* configuration in conditional on the shield symbol
 - [ ] `.conf` file has optional extra features commented out
 - [x] Keyboard/PCB is part of a shipped group buy or is generally available in stock to purchase (OSH/personal projects without general availability should create a zmk-config repo instead)

### CODE REVIEW

**Comment by caksoylar** [comment]
Is this definition intended to be used along with other shields using a build definition with multiple shields? If so, does the Github actions workflow support such a definition through `build.yaml`?

Is this also supposed to be compatible with nice!nano only (or with `pro micro` boards, for that matter)? I don't see why that should be, but it seems like the case from the SPI node alias.

**Comment by Nicell** [comment]
> Is this definition intended to be used along with other shields using a build definition with multiple shields? If so, does the Github actions workflow support such a definition through `build.yaml`?

Exactly! The GitHub actions workflow should work fine according to Pete... I need to test it though. Should just need to update the build.yaml shield prop to have `nice_view` after the existing one.

> Is this also supposed to be compatible with nice!nano only (or with `pro micro` boards, for that matter)? I don't see why that should be, but it seems like the case from the SPI node alias.

Should be compatible with all the other Pro Micro boards, I just need to define SPI nodes/aliases for others, which is on my todo.

File: app/boards/shields/nice_view/nice_view.conf [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Enable nice!view
+CONFIG_ZMK_DISPLAY=y
+CONFIG_SSD1306=n
```
Comment (petejohanson) [inline app/boards/shields/nice_view/nice_view.conf:3]: Why was this necessary? We used to have a bug or two with some random shield having conditionals wrong, causing this to auto-enable, but I don't think we still do?

File: app/boards/shields/nice_view/nice_view.overlay [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+		compatible = "sharp,ls0xx";
+		label = "DISPLAY";
+		spi-max-frequency = <1000000>;
+		reg = <0>;
+		width = <160>;
+		height = <68>;
+	};
+};
+
+&pro_micro_i2c {
```
Comment (petejohanson) [inline app/boards/shields/nice_view/nice_view.overlay:21]: This ties the shield to only working with PM based boards. Given XIAO gaining popularity, I wonder if this should be in `shields/nice_view/boards/*.overlay` files for all the existing PM boards?

File: app/boards/shields/nice_view/nice_view.zmk.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+file_format: "1"
+id: nice_view
+name: nice!view
+type: shield
+url: https://nicekeyboards.com/nice-view
+requires: [pro_micro, i2c_oled]
```
Comment (petejohanson) [inline app/boards/shields/nice_view/nice_view.zmk.yml:6]: We should consider this, this is an odd requirement, if we want this shield to be a bit more generic.

File: app/boards/shields/nice_view/nice_view.overlay [hunk 4]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2022 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+&nice_view_spi {
+	status = "okay";
+	cs-gpios = <&pro_micro 1 GPIO_ACTIVE_HIGH>;
```
Comment (petejohanson) [inline app/boards/shields/nice_view/nice_view.overlay:9]: I think this should be something like:

```suggestion
	cs-gpios = <&nice_view_d 0 GPIO_ACTIVE_HIGH>;
```

And then expect a `nice_view_d` nexus node to be supplied? Possibly with some helper macro to make it not so painful to write every time? Or with an include so you can just write:

````
&nice_view_d {
  gpio-map = <0 0 &pro_micro 1 0>;
};
```

As needed to add the mapping.

File: app/boards/shields/nice_view/nice_view.overlay [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2022 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+&nice_view_spi {
```
Comment (petejohanson) [inline app/boards/shields/nice_view/nice_view.overlay:7]: Should we provide this in `shields/nice_view/boards/*.overlay` files for common/in-tree boards?

File: app/boards/arm/nice_nano/nice_nano.dtsi [hunk 6]
```diff
@@ -63,6 +70,7 @@
 	};
 };
 
+nice_view_spi: &spi0 {};
```
Comment (petejohanson) [inline app/boards/arm/nice_nano/nice_nano.dtsi:73]: Move this to the boards override directory in the shield itself.

**Review by petejohanson — COMMENTED** [review]
Some initial thoughts!

File: app/boards/shields/nice_view/nice_view.conf (hunk 1 above)
Comment (Nicell) [reply app/boards/shields/nice_view/nice_view.conf:3]: This was necessary as setting the `default` to `n` in the defconfig did not actually disable it since it's turned on in the other shield when `ZMK_DISPLAY` is enabled.

**Comment by Nicell** [comment]
Our requires/exposes system doesn't really handle this new type of interconnect (built on top of a complete keyboard, not to create one). We'll need to figure out how we handle this with our build system and our hardware docs page. I've set up stubs and done nothing with them for now for both the build system and docs.

File: app/boards/shields/nice_view/Kconfig.defconfig [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+config LS0XX
+	default y
+
+if !ZMK_SPLIT || ZMK_SPLIT_ROLE_CENTRAL
+
+config ZMK_WIDGET_WPM_STATUS
+	default y
+
+endif # !ZMK_SPLIT || ZMK_SPLIT_ROLE_CENTRAL
```
Comment (petejohanson) [inline app/boards/shields/nice_view/Kconfig.defconfig:22]: Can't this be combined to a single default with condition?


```suggestion
config ZMK_WIDGET_WPM_STATUS
	default y if !ZMK_SPLIT || ZMK_SPLIT_ROLE_CENTRAL
```

File: app/boards/shields/nice_view/nice_view.zmk.yml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+file_format: "1"
+id: nice_view
+name: nice!view
+type: shield
+url: https://nicekeyboards.com/nice-view
+requires: [nice_view]
```
Comment (petejohanson) [inline app/boards/shields/nice_view/nice_view.zmk.yml:6]:
```suggestion
requires: [nice_view_header]
```

?

File: app/boards/shields/nice_view_oled/Kconfig.shield [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2022 The ZMK Contributors
+# SPDX-License-Identifier: MIT
+
+config SHIELD_NICE_VIEW_OLED
+	def_bool $(shields_list_contains,nice_view_oled)
```
Comment (petejohanson) [inline app/boards/shields/nice_view_oled/Kconfig.shield:5]: As I think about this... Still unsure about the name.

OLED...

- adapter?
- bodge
- converter

Some alternative ideas.

File: app/boards/shields/nice_view_oled/boards/nice_nano.overlay [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+ * SPDX-License-Identifier: MIT
+ */
+
+nice_view_spi: &spi0 {
+	compatible = "nordic,nrf-spim";
+	status = "okay";
+	sck-pin = <20>;
+	mosi-pin = <17>;
+	miso-pin = <25>;
+	cs-gpios = <&pro_micro 1 GPIO_ACTIVE_HIGH>;
```
Comment (petejohanson) [inline app/boards/shields/nice_view_oled/boards/nice_nano.overlay:13]: I think omit this entirely? Have folks override the `cs-gpios` property in their keymap for whatever pin they bodged.

**Review by petejohanson — CHANGES_REQUESTED** [review]
A few follow up thoughts.

I think each shield deserves a readme file as well to explain their use, lacking tooling for this in some setup script.

File: app/boards/shields/nice_view_oled/boards/nice_nano.overlay (hunk 10 above)
Comment (Nicell) [reply app/boards/shields/nice_view_oled/boards/nice_nano.overlay:13]: It'd be nice to allow for this to be a default fallback. I'll likely put in the docs for the nice!view to use this pin if available. (it should be for most since it's on the trrs serial pin)

Comment (petejohanson) [reply app/boards/shields/nice_view_oled/boards/nice_nano.overlay]: As long as this is clearly documented there, and in the README, that's fine.

**Review by petejohanson — CHANGES_REQUESTED** [review]
I think last thing: can you please add board overrides for the other in-tree nRF52 boards, e.g. nRFMicro, etc?

File: app/boards/shields/nice_view_adapter/README.md [hunk 11]
```diff
@@ -0,0 +1,11 @@
+# nice!view Adapter
+
+This shield is used as an adapter between the nice!view and existing shields/boards that expose an I2C OLED header.
+
+To use this shield, you should add this shield to your list of shields *before* `nice_view`.
+
+The nice!view will use the SDA/SCL pins of the OLED, and then the adapter expects a final pin to be "bodged" from your microcontroller to the nice!view CS pin. This adapter assumes that the CS pin bodged is the `&pro_micro 1 pin` or "D1", which is the top left pin when looking at the front of the board. If you can't use this pin, you'll need to write your own definition for the `&nice_view_spi` bus.
```
Comment (caksoylar) [inline app/boards/shields/nice_view_adapter/README.md:7]:
```suggestion
The nice!view will use the SDA/SCL pins of the header, and then the adapter expects a final pin to be "bodged" from your microcontroller to the nice!view CS pin. This adapter assumes that the CS pin bodged is the `&pro_micro 1 pin` or "D1", which is the top left pin when looking at the front of the board. If you can't use this pin, you'll need to override the `cs-gpios` definition for the `&nice_view_spi` bus.
```
The first sentence might be confusing because there is actually no OLED display involved here. Also thoughts on making it more specific as such for the last sentence?

Comment (Nicell) [reply app/boards/shields/nice_view_adapter/README.md:7]: Agreed on the first sentence. Second sentence... I think the override is only possible on zmk-config keymaps? The reason being is we need the override to be the last DT override, and I think currently that's only possible with the zmk-config if I'm not mistaken.

Comment (caksoylar) [reply app/boards/shields/nice_view_adapter/README.md:7]: I might be misunderstanding, but I believe you always need it to be defined last, right? Even if you create a new `nice_view_spi` node instead of overriding with `&nice_view_spi`, it still needs to be defined later for it to take effect instead of the node definition coming from board-specific overlays.

**Review by caksoylar — APPROVED** [review]

File: app/boards/shields/nice_view_adapter/nice_view_adapter.zmk.yml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+file_format: "1"
+id: nice_view_adapter
+name: nice!view
+type: shield
+url: https://nicekeyboards.com/nice-view
+requires: [i2c_oled]
+exposes: [nice_view_header]
+features:
+  - display
```
Comment (petejohanson) [inline app/boards/shields/nice_view_adapter/nice_view_adapter.zmk.yml:9]: I think this should be removed. The adapter itself doesn't add a display.

**Review by petejohanson — CHANGES_REQUESTED** [review]
Sorry, one last thing I noticed after a re-read of this.

File: app/boards/shields/nice_view_adapter/nice_view_adapter.zmk.yml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+file_format: "1"
+id: nice_view_adapter
+name: nice!view
```
Comment (petejohanson) [inline app/boards/shields/nice_view_adapter/nice_view_adapter.zmk.yml:3]:
```suggestion
name: nice!view adapter
```

**Review by petejohanson — CHANGES_REQUESTED** [review]

**Review by petejohanson — APPROVED** [review]
Thanks!
